src/main/java/org/openlmis/buq/web/buq/BottomUpQuantificationController.java

```diff
@@ -0,0 +1,216 @@
+
+    return buildDto(bottomUpQuantification);
+  }
+
+  /**
+   * Allows updating bottom-up quantification.
+   *
+   * @param bottomUpQuantificationId UUID of bottom-up quantification which we want to update.
+   * @param bottomUpQuantificationDto A bottom-up quantification DTO bound to the request body.
+   * @return updated bottom-up quantification.
+   */
+  @PutMapping("/{id}")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public BottomUpQuantificationDto save(@PathVariable("id") UUID bottomUpQuantificationId,
+      @RequestBody BottomUpQuantificationDto bottomUpQuantificationDto) {
+    if (!bottomUpQuantificationRepository.existsById(bottomUpQuantificationId)) {
+      throw new NotFoundException(MessageKeys.ERROR_BOTTOM_UP_QUANTIFICATION_NOT_FOUND);
+    }
+
+    BottomUpQuantification updatedBottomUpQuantification = bottomUpQuantificationService
+        .save(bottomUpQuantificationDto, bottomUpQuantificationId);
+
+    return buildDto(updatedBottomUpQuantification);
+  }
+
+
+  /**
+   * Retrieves audit information related to the specified bottom-up quantification.
+   *
+   * @param author The author of the changes which should be returned.
+   *               If null or empty, changes are returned regardless of author.
+   * @param changedPropertyName The name of the property about which changes should be returned.
+   *               If null or empty, changes associated with any and all properties are returned.
+   * @param page A Pageable object that allows client to optionally add "page" (page number)
+   *             and "size" (page size) query parameters to the request.
+   */
+  @GetMapping(value = "/{id}/auditLog")
+  @ResponseStatus(HttpStatus.OK)
+  @ResponseBody
+  public ResponseEntity<String> getBottomUpQuantificationAuditLog(@PathVariable("id") UUID id,
+      @RequestParam(name = "author", required = false, defaultValue = "") String author,
+      @RequestParam(name = "changedPropertyName", required = false, defaultValue = "")
+      String changedPropertyName, Pageable page) {
+
+    // Return a 404 if the specified instance can't be found
+    if (!bottomUpQuantificationRepository.existsById(id)) {
+      throw new NotFoundException(MessageKeys.ERROR_BOTTOM_UP_QUANTIFICATION_NOT_FOUND);
+    }
+
+    return getAuditLogResponse(BottomUpQuantification.class, id, author, changedPropertyName,
+        page);
+  }
+
+  private BottomUpQuantificationDto buildDto(BottomUpQuantification bottomUpQuantification) {
+    BottomUpQuantificationDto dto = new BottomUpQuantificationDto();
+    bottomUpQuantification.export(dto);
+
+    List<BottomUpQuantificationLineItem> bottomUpQuantificationLineItems =
+        bottomUpQuantification.getBottomUpQuantificationLineItems();
+    List<BottomUpQuantificationLineItemDto> lineItemDtoList = bottomUpQuantificationLineItems
+        .stream()
+        .map(lineItem -> {
+          BottomUpQuantificationLineItemDto lineItemDto = new BottomUpQuantificationLineItemDto();
```

Comment:
Hmm, aren't there any utility method where you would just go `BottomUpQuantificationLineItemDto.fromExport(lineItemDto)` or something - looks like a common use.
The other one is couple of lines below.

Reply:
I changed it for **map(BottomUpQuantificationLineItemDto::newInstance)**